### PR TITLE
added bysetpos validation to prevent endless looping

### PR DESCRIPTION
**Pull request by LittleJono**
TL;DR of the problem:

In an iCal feed you can define a recurring schedule and also say "pick every 3rd event of each month". However, this doesn't work if you also say "This event happens on the third sunday of each month". You suddenly have conflicting criteria, the worker starts looking for the third event of each month, when each month only has 1 event. Currently if the worker encounters this it will spend a very very very very very long time generating events for a month and then checking if any events of that month meet the criteria. It does have a hard-coded kill-loop value I believe but it isn't reached in a timely manner.

### CODE REVIEW

File: lib/ical/recur_iterator.js [hunk 1]
```diff
@@ -409,6 +419,39 @@ ICAL.RecurIterator = (function() {
       this.increment_generic(inc, "hour", 24, "monthday");
     },
 
+    get_number_of_days_in_frequency: function get_number_of_days_in_frequency() {
+      // There is a maximum of 53 of any given day in a year,
+      // and 5 in a month. i.e. there cannot be more than 5 Sundays in 1 month.
+      return this.rule.freq === "YEARLY" ? 53 : 5;
```
Comment (luisbosch) [inline lib/ical/recur_iterator.js]: Are there any "enums" for 'YEARLY' ?

File: test/performance/rrule_test.js [hunk 2]
```diff
@@ -43,7 +43,8 @@ perfCompareSuite('rrule', function(perf, ICAL) {
 
     // Apple iCal rules
     "FREQ=MONTHLY;BYDAY=MO,TU,WE,TH,FR;BYSETPOS=3",
-    "FREQ=MONTHLY;BYDAY=SA,SU;BYMONTH=11;BYSETPOS=-1"
+    "FREQ=MONTHLY;BYDAY=SA,SU;BYMONTH=11;BYSETPOS=-1",
+    "FREQ=MONTHLY;BYDAY=3MO,3TU;BYSETPOS=3"
```
Comment (luisbosch) [inline test/performance/rrule_test.js:47]: What is this rule for?

File: lib/ical/recur_iterator.js [hunk 3]
```diff
@@ -327,6 +327,16 @@ ICAL.RecurIterator = (function() {
         return null;
       }
 
+      // if there are no possible valid events return null.
+      try {
+        if (this.has_invalid_bysetpos()) {
+          return null;
+        }
+      } catch (err) {
+       // do nothing.
```
Comment (luisbosch) [inline lib/ical/recur_iterator.js]: does it make sense to continue if there was en exception thrown in `has_invalid_bysetpos` ?

**Comment by luisbosch** [comment]
Do the tests in `ical` cover what you've added? Should we add more tests?

File: test/performance/rrule_test.js (hunk 2 above)
Comment (LittleJono) [reply test/performance/rrule_test.js:47]: `FREQ=MONTHLY;BYDAY=3MO,3TU` returns an array of 2 events [third monday, third tuesday]. `BYSETPOS=3` attempts to grab the 3rd event from the array, which doesn't exist. Without this fix the parser will get stuck in an endless loop looking for the 3rd event. This rule tests that it doesn't get stuck in a loop, without the fix having this rule here makes the tests never return.

File: lib/ical/recur_iterator.js (hunk 3 above)
Comment (LittleJono) [reply lib/ical/recur_iterator.js]: That code should never throw an error, this is just to make sure I don't accidentally break the parser if there's an edge case that isn't covered by the unit tests.

**Comment by LittleJono** [comment]
> Do the tests in `ical` cover what you've added? Should we add more tests?

Yeah I need to add some more tests I think, just want to get some opinions first to make sure what I've got so far is decent.

File: lib/ical/recur_iterator.js (hunk 1 above)
Comment (luisbosch) [reply lib/ical/recur_iterator.js]: Nevermind, just saw it's used like that before.

File: test/performance/rrule_test.js (hunk 2 above)
Comment (luisbosch) [reply test/performance/rrule_test.js:47]: Ok, it's just testing data. This was the one causing errors?

Comment (LittleJono) [reply test/performance/rrule_test.js:47]: yep

**Review by luisbosch — APPROVED** [review]
Looks good. But you may want to get another pair of eyes, more familiar with iCal.

File: lib/ical/recur_iterator.js (hunk 1 above)
Comment (andy-haynes) [inline lib/ical/recur_iterator.js]: is this always either `YEARLY` or `MONTHLY` then?

File: lib/ical/recur_iterator.js [hunk 4]
```diff
@@ -409,6 +419,39 @@ ICAL.RecurIterator = (function() {
...
+      }
+
+      if (!this.by_data.BYSETPOS || this.by_data.BYSETPOS.length < 1) {
+        return false;
+      }
+
+      var max_number_of_possible_days = this.get_max_number_of_possible_byday_values();
+      var checker = function(acc, value) {
+        if (acc) {
+          return acc;
```
Comment (andy-haynes) [inline lib/ical/recur_iterator.js]: This `reduce` checks every value in `by_data.BYSETPOS` even if it finds an invalid entry. It might be worth writing a loop that returns early instead to further minimize iteration.

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Seems sane to me, just a couple questions inline. I'm no ical expert but so long as we can run comparisons for affected and non-affected feeds in MAS I think this is viable.

File: lib/ical/recur_iterator.js (hunk 1 above)
Comment (LittleJono) [reply lib/ical/recur_iterator.js]: This should always be YEARLY or MONTHLY but DAILY and WEEKLY are technically also valid values even though it makes no sense to use them here, we should probably provide support for them regardless, I'll add it in

Comment (LittleJono) [reply lib/ical/recur_iterator.js]: I've come up with a better way of doing this, take a look.

File: lib/ical/recur_iterator.js (hunk 4 above)
Comment (LittleJono) [reply lib/ical/recur_iterator.js]: wish there was lodash here :P

**Comment by LittleJono** [comment]
A good site for validating all the various rules: https://jakubroztocil.github.io/rrule/

File: test/recur_iterator_test.js [hunk 5]
```diff
@@ -226,6 +227,38 @@ suite('recur_iterator', function() {
       assert.deepEqual(dates, options.dates || []);
     });
   }
+
+  function testRRULEWithError(ruleString, options, expectedError) {
```
Comment (andy-haynes) [inline test/recur_iterator_test.js:230]: It doesn't look like `expectedError` gets a value, I'm assuming if these tests pass then `assert.throws(...)` only takes two parameters?

File: lib/ical/recur_iterator.js [hunk 6]
```diff
@@ -409,6 +421,49 @@ ICAL.RecurIterator = (function() {
...
+        if (value > max_number_of_possible_days) {
+          return true;
+        }
+      }
+      return false;
+    },
+
+    has_invalid_bysetpos: function has_invalid_bysetpos(days_in_timeslice) {
+      // days_in_timeslice math:
+      // there are at most 5 of any given day in a month, and 53 of any given day in a year
```
Comment (andy-haynes) [inline lib/ical/recur_iterator.js:449]: Might be worth duplicating the bit about 5 days in a month where the `5` appears above to keep it from being mistaken for a magic number.

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Looking really good (not that I can read ical formatting 😛)!

Just a couple minor comments but otherwise LGTM 🎉

**Review by andy-haynes — APPROVED** [review]
LGTM 🌮
